### ISSUE TITLE
Allow for serving GET operation from cache if user explicitly opt-in for it

### ISSUE THREAD

**Issue by wojtek-t** [kubernetes/kubernetes#37473]
Currently, if you are sending "LIST" request to apiserver, depending on the "ResourceVersion" field value:
https://github.com/kubernetes/kubernetes/blob/8ef325643490274f8639ea1bee8c7718b316e839/pkg/api/v1/types.go#L3247
the request will be served:
- from etcd if it is not set
- from apiserver cache, if it is set with a guarantee that the returned result is at least as fresh as the "RV" passed as parameter.
Serving request from etcd is a default behavior, but if client can tolerate a bit unfresh data (e.g. if the list is just a starting point for a watch), then it can opt-in for listing from apiserver cache. This mechanism is used e.g. in the reflector/informer framework and is significant optimization for the system.

We should allow for exactly the same for "GET" operations.

@lavalamp @kubernetes/sig-api-machinery

**Comment by lavalamp** [comment]
👍 

We need to make sure this is documented in the OpenAPI spec.

@bgrant0607 We were talking about this earlier today.

**Comment by hongchaodeng** [comment]
@wojtek-t @lavalamp 
Mind share if you know any use cases that would benefit from this?
Do you know how much faster it would be?

**Comment by wojtek-t** [comment]
Yup - this is hack to that we can get rid of once we have it:
https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kubelet_node_status.go#L312

This speeded up 99th % of GET node operations by ~10x and speeded up pretty much all other operations by reducing load on etcd. There are at least few other places we can leaverage it (e.g. NodeProblemDetector).

**Comment by bgrant0607** [comment]
@wojtek-t Once this is introduced, it will be subject to API deprecation policy.

**Comment by xiang90** [comment]
@wojtek-t This sounds a correct thing to do.

**Comment by liggitt** [comment]
Exposing semantics of waiting for a (potentially far) future version in this way are odd to me (get resourceVersion=8 may not return resourceVersion 8 of that resource). I guess something similar was already done for list (though I saw more discussion of "serve from cache" aspect) but this is doubling down… how are we using the blocking aspect of the list resourceVersion? are we happy with that mechanism applied to a single resource that may not match the resourceVersion?

Also, by making API users opt into "serve from API server cache", what are the implications if they don't? Reads from etcd are already not guaranteed to be quorum reads… will we need a "force quorum read" option in the future alongside this?

**Comment by lavalamp** [comment]
No-- all of our config turns on quorum reads. (If it doesn't, it's a bug.)
Clients need to explicitly opt out of quorum read only if they know what
they're doing.

**Comment by smarterclayton** [comment]
I'm fairly worried about long waits.  It's convenient, and makes certain
clients easier to write.  But it is a big semantic change and I'm not
comfortable backing into it.

On Dec 5, 2016, at 7:34 PM, Jordan Liggitt <notifications@github.com> wrote:

Exposing semantics of waiting for a (potentially far) future version in
this way are odd to me (get resourceVersion=8 may not return
resourceVersion 8 of that resource). I guess something similar was already
done for list (though I saw more discussion of "serve from cache" aspect)
but this is doubling down… how are we using the blocking aspect of the list
resourceVersion? are we happy with that mechanism applied to a single
resource that may not match the resourceVersion?

Also, by making API users opt into "serve from API server cache", what are
the implications if they don't? Reads from etcd are already not guaranteed
to be quorum reads… will we need a "force quorum read" option in the future
alongside this?

—
You are receiving this because you are on a team that was mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/kubernetes/kubernetes/issues/37473#issuecomment-265024408>,
or mute the thread

**Comment by wojtek-t** [comment]
@smarterclayton I don't think it's a semantic change. We are introducing a new option for which you need to explicitly opt-in. The default behavior remains as it was before.

**Comment by smarterclayton** [comment]
The wait is a semantic change to how the server works.  I'm not really
happy with unbounded GET request waits being added without clearly thinking
through the implications on denial of service on the cluster.

For example, what if I say resourceVersion=MaxInt64?  And then fill up all
the rate limiter?

On Tue, Dec 6, 2016 at 3:45 AM, Wojciech Tyczynski <notifications@github.com

**Comment by lavalamp** [comment]
Wait, it waits in that situation? I thought it just did a quorum read if
the requested rv was greater than what was in the cache.

**Comment by liggitt** [comment]
actually, I'm really confused what resourceVersion is currently doing with list:

```
# no resource version returns immediately
$ curl  -H "Accept: application/yaml" http://localhost:8080/api/v1/namespaces/default/serviceaccounts
apiVersion: v1
items:
- metadata:
    creationTimestamp: 2016-12-07T03:22:49Z
    name: default
    namespace: default
    resourceVersion: "63"
    selfLink: /api/v1/namespaces/default/serviceaccounts/default
    uid: 6e59d7f6-bc2c-11e6-8388-acbc32c1ca87
  secrets:
  - name: default-token-3glxm
kind: ServiceAccountList
metadata:
  resourceVersion: "1578"
  selfLink: /api/v1/namespaces/default/serviceaccounts

# resource version matching item in list returns immediately
$ curl -H "Accept: application/yaml" http://localhost:8080/api/v1/namespaces/default/serviceaccounts?resourceVersion=63
apiVersion: v1
items:
- metadata:
    creationTimestamp: 2016-12-07T03:22:49Z
    name: default
    namespace: default
    resourceVersion: "63"
    selfLink: /api/v1/namespaces/default/serviceaccounts/default
    uid: 6e59d7f6-bc2c-11e6-8388-acbc32c1ca87
  secrets:
  - name: default-token-3glxm
kind: ServiceAccountList
metadata:
  resourceVersion: "64"
  selfLink: /api/v1/namespaces/default/serviceaccounts

# resource version matching item in list + 1 returns immediately
$ curl -H "Accept: application/yaml" http://localhost:8080/api/v1/namespaces/default/serviceaccounts?resourceVersion=64
apiVersion: v1
items:
- metadata:
    creationTimestamp: 2016-12-07T03:22:49Z
    name: default
    namespace: default
    resourceVersion: "63"
    selfLink: /api/v1/namespaces/default/serviceaccounts/default
    uid: 6e59d7f6-bc2c-11e6-8388-acbc32c1ca87
  secrets:
  - name: default-token-3glxm
kind: ServiceAccountList
metadata:
  resourceVersion: "64"
  selfLink: /api/v1/namespaces/default/serviceaccounts

# resource version matching item in list + 2 hangs forever
$ curl -H "Accept: application/yaml" http://localhost:8080/api/v1/namespaces/default/serviceaccounts?resourceVersion=65
```

**Comment by wojtek-t** [comment]
I think there might be some bugs in the implementation, as we are pretty much using only "resourceVersion=0". So we definitely need to look into this hanging forever issues - that definitely looks like a bug.

**Comment by wojtek-t** [comment]
@liggitt OK - so I finally got to it, and most of it is working as intended.

Here is the explanation (as I could easily reproduce it).
- there were 4 actions in the cluster touching ServiceAccounts:
1. create SA for default namespace
2. modify SA for default namespace
3. create SA for kube-system namespace
4. modify SA for kube-system namespace
- in your usecase you are only listing SA from default namespace, so we ignore 3 and 4
- in cache we have all of those events, and the largest RV from those is 64
- however, the larger from default namespace is 63

The contract of "resourceVersion" passed to List is:
```
The returned result is at least as fresh as the passed resourceVersion.
```

So since the cache is at rv=64, it returns rv=64 (for both 63 and 64 params).

The reason why we hang with rv=65 parameter is that cache is per resource, and there weren't any further events after that. Thus 65 is higher that whatever we observed so far.

I've just sent out #38705 out for review to reduce timeout (before it was waiting for 1 minute, in the PR I'm returning "BadRequest" after 3 seconds).

**Comment by liggitt** [comment]
it seems odd that the list resourceVersion initially received for http://localhost:8080/api/v1/namespaces/default/serviceaccounts (1578) results in a hang or BadRequest when requested like http://localhost:8080/api/v1/namespaces/default/serviceaccounts?resourceVersion=1578... the cache is that fresh

**Comment by liggitt** [comment]
partially implemented in 1.6, work remaining to complete in 1.7

**Comment by lavalamp** [comment]
is this done?

**Comment by liggitt** [comment]
xref https://github.com/kubernetes/kubernetes/issues/59848
